java/vector/src/main/java/org/apache/arrow/vector/BaseFixedWidthVector.java

```diff
@@ -870,20 +875,38 @@ public boolean equals(int index, ValueVector to, int toIndex) {
     if (to == null) {
       return false;
     }
-    if (this.getClass() != to.getClass()) {
+    if (!this.getField().getType().equals(to.getField().getType())) {
       return false;
     }
 
     BaseFixedWidthVector that = (BaseFixedWidthVector) to;
 
-    int leftStart = typeWidth * index;
-    int leftEnd = typeWidth * (index + 1);
+    boolean isNull = isNull(index);
+    if (isNull != that.isNull(toIndex)) {
+      return false;
+    }
+
+    if (!isNull) {
+      if (this.isSet(index) != that.isSet(toIndex)) {
+        return false;
+      }
+
+      int leftStart = typeWidth * index;
+      int leftEnd = typeWidth * (index + 1);
+
+      int rightStart = typeWidth * toIndex;
+      int rightEnd = typeWidth * (toIndex + 1);
+
+      int ret = ByteFunctionHelpers.equal(this.getDataBuffer(), leftStart, leftEnd,
+          that.getDataBuffer(), rightStart, rightEnd);
+      return ret == 1;
+    }
 
-    int rightStart = typeWidth * toIndex;
-    int rightEnd = typeWidth * (toIndex + 1);
+    return true;
+  }
 
-    int ret = ByteFunctionHelpers.equal(this.getDataBuffer(), leftStart, leftEnd,
-        that.getDataBuffer(), rightStart, rightEnd);
-    return ret == 1;
+  @Override
+  public boolean accept(RangeEqualsVisitor visitor) {
```

Comment:
Equals(int index, ValueVector vector, int toIndex) compares single value and RangeEqualsVisitor compares a range of values.

Reply:
Yes, we did a refactor for dictionary encoding few weeks ago, hashCode/equals API are used for DictionaryHashTable (https://github.com/apache/arrow/blob/master/java/vector/src/main/java/org/apache/arrow/vector/dictionary/DictionaryHashTable.java#L138) to avoid memory copy introduced by previous implementation.

Reply:
agree with @BryanCutler - this causes duplication of code/tests. Is there a reason to not use the RangeEquals API (with range = 1) instead ?

I'm fine if you want to do this in a different PR.

Reply:
@pravindra Ah, I have already made it reuse RangeEquals API(range ==1).

Reply:
Sorry, I meant can we remove the equals() API from ValueVector, and it's callers be modified to instead use the RangeEquals() API ? I think keeping the minimal interface in ValueVector is ideal.

Reply:
I see, right, could be done in a different PR.